src/libraries/Common/src/System/Net/Mail/DomainLiteralReader.cs

```diff
@@ -25,43 +25,62 @@ internal static class DomainLiteralReader
         // - The next index past the terminating bracket (data[index + 1] == StartSquareBracket).
         //   e.g. In (user@[domain]), starting at index=12 (]) returns index=4 (@).
         //
-        // A FormatException will be thrown if:
+        // A FormatException will be thrown or false is returned if:
         // - A non-escaped character is encountered that is not valid in a domain literal, including Unicode.
         // - The final bracket is not found.
         //
-        internal static int ReadReverse(string data, int index)
+        internal static bool TryReadReverse(string data, int index, out int outIndex, bool throwExceptionIfFail)
         {
             Debug.Assert(0 <= index && index < data.Length, "index was outside the bounds of the string: " + index);
             Debug.Assert(data[index] == MailBnfHelper.EndSquareBracket, "data did not end with a square bracket");
 
+            outIndex = default;
```

Comment:
I generally don't like code that does this because it defeats the helpful compile-time correctness checking. It also creates a stray write in the event of success, though that is probably going to be swallowed by the rest of the code here.

Reply:
@stephentoub do we have any guidance on this?

Reply:
You mean use `default(int)` and put it only on false branches?

Reply:
I don't care about `default` vs `default(int)`, more just doing `out = default` at the start of a method. It's convenient because it lets you avoid writing an extra line on the fail path but it also defeats the "was an out param assigned" check.

Reply:
In general we try not to do this, i.e. the guidance is "store the value only once you have a result".  In some cases, though, either because there are lots of exit points and it's not worth all of the repetitions / code increase, or if there's a safety aspect of ensuring nothing stale gets left, we'll do the up-front write.

Reply:
Oh sorry Cory I misunderstood, clear I'll amend.